source/game/field/obj/ObjectTwistedWay.cc

```diff
@@ -0,0 +1,355 @@
+        }
+
+        if (pInfo) {
+            pInfo->bbox.min = pInfo->bbox.min.minimize(bbox);
+            pInfo->bbox.max = pInfo->bbox.max.maximize(bbox);
+
+            if (full) {
+                pInfo->updateWall(dist, wnrm);
+            }
+        }
+
+        if (pFlagsOut) {
+            if (push) {
+                CollisionDirector::Instance()->pushCollisionEntry(dist, pFlagsOut,
+                        KCL_TYPE_BIT(COL_TYPE_WALL), 0xc);
+            } else {
+                *pFlagsOut |= KCL_TYPE_BIT(COL_TYPE_WALL);
+            }
+        }
+
+        return true;
+    }
+}
+
+bool ObjectTwistedWay::checkSphereRoadCollision(f32 heightOffset, f32 radius,
+        const EGG::Vector3f &relativePos, CollisionInfo *pInfo, KCLTypeMask *pFlagsOut, bool full,
+        bool push) {
+    auto [sinfidx, cosfidx] = EGG::Mathf::SinCosFIdx(heightOffset * RAD2FIDX);
+
+    f32 scaledB8 = _B8 * -0.5f;
+    f32 dist = scaledB8 + radius + -relativePos.x * sinfidx + cosfidx * (-relativePos.y - scaledB8);
+
+    if (dist <= 0.0f) {
+        return false;
+    }
+
+    EGG::Vector3f bbox = dist * EGG::Vector3f::ey;
+    EGG::Vector3f fnrm = EGG::Vector3f(sinfidx, cosfidx, 0.0f);
+
+    if (pInfo) {
+        pInfo->bbox.min = pInfo->bbox.min.minimize(bbox);
+        pInfo->bbox.max = pInfo->bbox.max.maximize(bbox);
+
+        if (full) {
+            pInfo->updateFloor(dist, fnrm);
+        }
+    }
+
+    if (pFlagsOut) {
+        if (push) {
+            auto *colDirector = CollisionDirector::Instance();
+            colDirector->pushCollisionEntry(dist, pFlagsOut, KCL_TYPE_BIT(COL_TYPE_ROAD), 0x0);
+            colDirector->setCurrentCollisionVariant(2);
+            if (EGG::Mathf::abs(relativePos.z) < COLLISION_RADIUS * TRICKABLE_RADIUS_FACTOR) {
+                colDirector->setCurrentCollisionTrickable(true);
+            }
+        } else {
+            *pFlagsOut |= KCL_TYPE_BIT(COL_TYPE_ROAD);
+        }
+    }
+
+    return true;
+}
+
+/// @brief Helper function which contains frequently re-used code. Behavior branches depending on
+/// whether it is a full or partial check (call CollisionInfo::updateFloor) or push (push entry in
+/// the CollisionDirector).
+/// @param full Whether to update the floor in the CollisionInfo
+/// @param push Whether to push a collision entry
+bool ObjectTwistedWay::checkSphereImpl(f32 radius, const EGG::Vector3f &v0,
+        const EGG::Vector3f & /*v1*/, KCLTypeMask flags, CollisionInfo *pInfo,
+        KCLTypeMask *pFlagsOut, u32 timeOffset, bool full, bool push) {
+    EGG::Vector3f relativePos = v0 - m_pos;
+
+    if (EGG::Mathf::abs(relativePos.z) > COLLISION_RADIUS ||
+            EGG::Mathf::abs(relativePos.x) > WIDTH * 2.0f) {
+        return false;
+    }
+
+    u32 frameCount;
+    if (!System::RaceManager::Instance()->isStageReached(System::RaceManager::Stage::Race)) {
+        frameCount = m_ac;
+    } else {
+        frameCount = System::RaceManager::Instance()->timer();
+    }
```

Comment:
Maybe overkill with the ternary. More importantly, we should only get the pointer here once. Writing the boolean out to a variable avoids recomputing it again in your next if-statement
```suggestion
    auto *raceMgr = System::RaceManager::Instance();
    bool raceStarted = raceMgr->isStageReached(System::RaceManager::Stage::Race);
    m_frameCount = raceStarted ? raceMgr->timer() : m_ac;
```

Reply:
I've seen other cases in Kinoko where an instance getter is called multiple times in a function, for instance CollisionDirector gets re-fetched multiple times per iteration in a couple for loops in KartMove and KartCollide, and BoxColManager gets re-fetched multiple times in a while loop in ObjectDirector::checkKartObjectCollision. Maybe cleaning those up could be a PR?